src/Grpc.AspNetCore.Server/Internal/GrpcServiceBinder.cs

```diff
@@ -18,47 +18,98 @@
 using Microsoft.AspNetCore.Routing;
 
 namespace Grpc.AspNetCore.Server.Internal
 {
     internal class GrpcServiceBinder<TService> : ServiceBinderBase where TService : class
     {
         private readonly IEndpointRouteBuilder _builder;
+        private readonly ServiceMethodsRegistery _serviceMethodsRegistery;
         private readonly ServerCallHandlerFactory<TService> _serverCallHandlerFactory;
 
         internal IList<IEndpointConventionBuilder> EndpointConventionBuilders { get; } = new List<IEndpointConventionBuilder>();
+        internal IList<IMethod> ServiceMethods { get; } = new List<IMethod>();
 
-        internal GrpcServiceBinder(IEndpointRouteBuilder builder, ServerCallHandlerFactory<TService> serverCallHandlerFactory)
+        internal GrpcServiceBinder(IEndpointRouteBuilder builder, ServiceMethodsRegistery serviceMethodsRegistery, ServerCallHandlerFactory<TService> serverCallHandlerFactory)
         {
             _builder = builder;
+            _serviceMethodsRegistery = serviceMethodsRegistery;
             _serverCallHandlerFactory = serverCallHandlerFactory;
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ClientStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateClientStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, DuplexStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateDuplexStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ServerStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateServerStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, UnaryServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateUnary(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
+        }
+
+        private void AddMethodCore(IMethod method, RequestDelegate requestDelegate)
+        {
+            ServiceMethods.Add(method);
```

Comment:
qq: what happens when we try to add a handler for the same method twice? (two "different" generated services that have the same URL, or registering the same service twice by mistake).
In gRPC C# registering the same method url twice is an error: https://github.com/grpc/grpc/blob/68381892d3fc52594e963a12b077e176021b36c4/src/csharp/Grpc.Core/Server.cs#L262

Reply:
At application startup nothing will happen. When someone attempts to go to a URL registered twice they will get an error that multiple endpoints match that URL.